vql/server/humio/humio.go

```diff
@@ -0,0 +1,707 @@
+
+	gMaxPoll = 60
+	gMaxPollDev = 30
+	gStatsLogPeriod = time.Duration(30) * time.Second
+	gNextId int32 = 0
+	gHttpClientTimeout = 10
+
+	apiEndpoint = "/v1/ingest/humio-structured"
+)
+
+type errInvalidArgument struct {
+	Arg string
+	Err error
+}
+
+func (err errInvalidArgument) Error() string {
+	return fmt.Sprintf("Invalid Argument - %s", err.Err)
+}
+
+func (err errInvalidArgument) Is(other error) bool {
+	return errors.Is(err.Err, other)
+}
+
+type errHttpClientError struct {
+	StatusCode int
+	Status string
+}
+
+func (err errHttpClientError) Error() string {
+	return err.Status
+}
+
+type errMaxRetriesExceeded struct {
+	LastError error
+	Retries int
+}
+
+func (err errMaxRetriesExceeded) Error() string {
+	return fmt.Sprintf("Maximum retries exceeded: %v attempts, last error=%s",
+			   err.Retries, err.LastError)
+}
+
+func (err errMaxRetriesExceeded) Unwrap() error {
+	return err.LastError
+}
+
+var errQueueOpened = errors.New("Cannot modify parameters of open queue")
+
+type HumioQueue struct {
+	scope                  	  vfilter.Scope
+	config                 	  *config_proto.Config
+
+	listener                  *directory.Listener
+	workerGrp                 *errgroup.Group
+	workerCtx		  context.Context
+
+	httpClient                *http.Client
+	// (bool) whether the queue is open
+	queueOpened		  int32
+
+
+	endpointUrl               string
+	authToken		  string
+	nWorkers                  int
+	tagMap                    map[string]string
+	batchingTimeoutDuration   time.Duration
+	httpClientTimeoutDuration time.Duration
+	eventBatchSize            int
+	httpTimeout		  int
+	debug			  bool
+	debugEventsEnabled	  bool
+	debugEventsMap		  map[int][]func(int)
+	id			  int
+	logPrefix		  string
+	maxRetries		  int
+
+	// Statistics
+	// count of events queued for posting across all workers
+	currentQueueDepth	  int32
+	// count of events dropped during shutdown across all workers
+	droppedEvents		  int32
+	// count of events successfully posted
+	postedEvents		  int32
+	// count of events that failed to post
+	failedEvents		  int32
+	// count of retries since startup
+	totalRetries		  int32
+
+}
+
+type HumioEvent struct {
+	// Could be string or int.  We don't need to care which.
+	Timestamp interface{}        `json:"timestamp"`
+	Attributes *ordereddict.Dict `json:"attributes"`
+	Timezone string              `json:"timezone,omitempty"`
+}
+
+type HumioPayload struct {
+	Events []HumioEvent          `json:"events"`
+	Tags map[string]interface{}  `json:"tags,omitempty"`
+}
+
+func (self *HumioPayload) String() string {
+	data, err := json.Marshal(self)
+	if err != nil {
+		return fmt.Sprintf("HumioPayload{unprintable, err=%s}", err)
+	}
+	
+	return string(data)
+}
+
+// This allows connection pooling across invocations
+func getHttpTransport() *http.Transport {
+	gHttpTransportLock.Lock()
+	defer gHttpTransportLock.Unlock()
+
+	if gHttpTransport == nil {
+		t := http.DefaultTransport.(*http.Transport).Clone()
```

Comment:
Sure. I think I'd like to add a GetHttpTransportWithTimeout, but otherwise it's fine. I developed this on 0.6.7-5 and forward ported it, and that didn't exist there.

Reply:
nm, i'm defining the timeout in the http.Client so that'll work fine